src/network/maybe_inf_delay.rs

```diff
@@ -0,0 +1,96 @@
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use futures::FutureExt;
+use futures_timer::Delay;
+use std::{
+	future::Future,
+	pin::Pin,
+	task::{Context, Poll, Waker},
+	time::Duration,
+};
+
+enum Inner {
+	Infinite {
+		/// Waker from the most recent `poll` call. If `None`, either `poll` has not been called
+		/// yet, we returned `Poll::Ready` from the last call, or the waker is attached to `delay`.
+		waker: Option<Waker>,
+		delay: Option<Delay>,
+	},
+	Finite(Delay),
+}
+
+pub struct MaybeInfDelay(Inner);
+
+impl MaybeInfDelay {
+	pub fn new(duration: Option<Duration>) -> Self {
+		match duration {
+			Some(duration) => Self(Inner::Finite(Delay::new(duration))),
+			None => Self(Inner::Infinite { waker: None, delay: None }),
+		}
+	}
+
+	pub fn reset(&mut self, duration: Option<Duration>) {
+		match duration {
+			Some(duration) => match &mut self.0 {
+				Inner::Infinite { waker, delay } => {
+					let mut delay = match delay.take() {
+						Some(mut delay) => {
+							delay.reset(duration);
+							delay
+						},
+						None => Delay::new(duration),
+					};
+					if let Some(waker) = waker.take() {
+						let mut cx = Context::from_waker(&waker);
+						match delay.poll_unpin(&mut cx) {
+							Poll::Pending => (), // Waker attached to delay
+							Poll::Ready(_) => waker.wake(),
+						}
+					}
+					self.0 = Inner::Finite(delay);
+				},
+				Inner::Finite(delay) => delay.reset(duration),
+			},
+			None =>
+				self.0 = match std::mem::replace(
+					&mut self.0,
+					Inner::Infinite { waker: None, delay: None },
+				) {
+					Inner::Finite(delay) => Inner::Infinite { waker: None, delay: Some(delay) },
+					infinite => infinite,
+				},
+		}
+	}
+}
+
+impl Future for MaybeInfDelay {
+	type Output = ();
+
+	fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+		match &mut self.0 {
+			Inner::Infinite { waker, .. } => {
+				*waker = Some(cx.waker().clone());
+				Poll::Pending
```

Comment:
Why not simply returning a Poll::Ready(())? then no need to handle the waker.
I mean if the point is to suspend until call to `reset` why not just send a message in a queue and get the poll in a real pending state by waiting on the queue (internally it is probably the same, just to simplify the code).

Reply:
Because `Poll::Ready` means the delay has elapsed, which should never be the case for an infinite delay. I'm not sure what you mean by sending a message in a queue? I'm also not sure in what way this is not a "real" pending state -- we return `Pending` and stash a clone of the waker, which is what every other `poll` implementation should be doing when not ready.

Reply:
you would have both an mpsc  receiver and a delay in this poll.
you only put it in pending when the receiver return pending and the delay return pending (or the delay is disabled somehow).
Then the thread is wake up by either a wake up from delay or a wakeup by sending a message (which would take place in your reset function).

Reply:
Oh I see. That sounds less inefficient than the current implementation? An MPSC queue is quite heavyweight compared to a single `Waker`. Also, currently I don't necessarily wake the polling thread in `reset`, just transfer the waker to the `Delay`.

Reply:
yes less efficient, not sure it matters, just a way to hide the waker.
Alternatively could just put the Delay to some very high value when we switch to Infinite and change value on reset (less efficient also).

Reply:
Really was not thinking straight yesterday, just reread and looks fine.
Reading a bit more I start to see the logic.
I wonder if it would be easier to use a single delay and make it always the closest item in queue and queried the queue eagerly (pop should only return if it deadline is correct given current time). Then if both queue return None, change delay to the delay from closest item in one of the queue, and if no item in queue then set the waker and return pending (enqueing item will always try to consume the waker).

Reply:
I did consider this but I didn't think the simpler API was worth the extra complexity underneath. Maybe I'll have a go at implementing it though and see how it turns out.